wal/src/message_queue_impl/region_meta.rs

```diff
@@ -0,0 +1,595 @@
+    /// sequences.
+    ///
+    /// NOTICE: Need to freeze the whole region meta on high-level before
+    /// calling.
+    pub async fn get_snapshot(&self) -> Result<RegionMetaSnapshot> {
+        let inner = self.inner.read().await;
+        // Calc the min offset in message queue.
+        let mut entries = Vec::with_capacity(inner.table_metas.len());
+        for (table_id, table_meta) in &inner.table_metas {
+            let meta_data = table_meta.get_meta_data().await?;
+            entries.push(meta_data);
+        }
+
+        Ok(RegionMetaSnapshot {
+            next_offset: inner.local_latest_offset.load(Ordering::Relaxed),
+            entries,
+        })
+    }
+}
+
+/// Region meta data.
+#[derive(Default, Debug)]
+struct RegionMetaInner {
+    table_metas: HashMap<TableId, TableMeta>,
+    /// It will fall behind the high watermark for ensuring successive
+    /// increment.
+    local_latest_offset: AtomicI64,
+}
+
+/// Wrapper for the [TableMetaInner].
+// TODO: will be made use of later.
+#[allow(unused)]
+#[derive(Debug)]
+struct TableMeta {
+    table_id: TableId,
+    /// The race condition may occur between writer thread
+    /// and background flush tread.
+    inner: Mutex<TableMetaInner>,
+}
+
+// TODO: will be made use of later.
+#[allow(unused)]
+impl TableMeta {
+    fn new(table_id: TableId) -> Self {
+        Self {
+            table_id,
+            inner: Mutex::new(TableMetaInner::default()),
+        }
+    }
+
+    async fn prepare_for_write(&self) -> SequenceNumber {
+        let inner = self.inner.lock().await;
+        inner.next_sequence_num
+    }
+
+    async fn update_after_write(&self, write_offset_range: (Offset, Offset), updated_num: u64) {
+        let mut inner = self.inner.lock().await;
+        let old_next_sequence_num = inner.next_sequence_num;
+        inner.next_sequence_num += updated_num;
+
+        // Update the mapping and high water mark.
+        let _ = inner
+            .start_sequence_offset_mapping
+            .insert(old_next_sequence_num, write_offset_range.0);
+        inner.current_high_watermark = write_offset_range.1 + 1;
+    }
+
+    async fn mark_deleted(&self, latest_marked_deleted: SequenceNumber) {
+        let mut inner = self.inner.lock().await;
+        assert!(latest_marked_deleted <= inner.next_sequence_num);
+        inner.latest_marked_deleted = latest_marked_deleted;
+
+        // Update the mapping, keep the range in description.
+        inner
+            .start_sequence_offset_mapping
+            .retain(|k, _| k >= &latest_marked_deleted);
+    }
+
+    async fn get_meta_data(&self) -> Result<TableMetaData> {
+        let inner = self.inner.lock().await;
+
+        // Only two situations exist:
+        // + no log of the table has ever been written
+        //  (next sequence num == latest marked deleted).
+        // + all logs of the table can be deleted
+        //  (next_sequence_num > latest_marked_deleted).
```

Comment:
This comment seems confusing to me -- Why are all the logs of table able to be deleted when `next_sequence_num > latest_marked_deleted`.